tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp2723Tests.cs

```diff
@@ -0,0 +1,286 @@
+
+                database.DropCollection("parents");
+                database.DropCollection("children");
+                database.DropCollection("grandChildren");
+
+                parents.InsertMany(new[]
+                {
+                    new Parent { Id = 1, Name = "parent1" },
+                    new Parent { Id = 2, Name = "parent2" }
+                });
+
+                children.InsertMany(new[]
+                {
+                    new Child { Id = 1, Name = "child1", ParentId = 2 },
+                    new Child { Id = 2, Name = "child2", ParentId = 1 }
+                });
+
+                grandChildren.InsertMany(new[]
+                {
+                    new GrandChild { Id = 1, Name = "grandchild1", ChildId = 2 },
+                    new GrandChild { Id = 2, Name = "grandchild2", ChildId = 1 }
+                });
+
+                var aggregate = parents
+                    .Aggregate()
+                    .Lookup<Parent, Child, ParentProjection>(
+                        children,
+                        parent => parent.Id,
+                        child => child.ParentId,
+                        parentProjection => parentProjection.Children)
+                    .Lookup<ParentProjection, GrandChild, GrandChild, IEnumerable<GrandChild>, ParentProjection>(
+                        grandChildren,
+                        let: new BsonDocument { { "children", "$Children" } },
+                        lookupPipeline: new BsonDocumentStagePipelineDefinition<GrandChild, GrandChild>(
+                            new[] { BsonDocument.Parse(@"{ $match : { $expr : { $and : [ { $in : [ ""$ChildId"", ""$$children._id"" ] } ] } } }") }),
+                        childProjection => childProjection.GrandChildren)
+                    .Project(parent => new ParentTree
+                    {
+                        Id = parent.Id,
+                        ParentName = parent.Name,
+                        Children = parent.Children
+                            .Select(child => new ChildTree
+                            {
+                                Id = child.Id,
+                                Name = child.Name,
+                                GrandChildren = parent
+                                    .GrandChildren
+                                    .Where(gc =>
+                                        gc.ChildId ==
+                                        child.Id)
+                                    .Select(gc =>
+                                        new GrandChildProjection()
+                                        {
+                                            Id = gc.Id,
+                                            Name = gc.Name
+                                        })
+                            })
+                    });
+
+                var stages = RenderStages(parents, aggregate);
+                var expectedStages = new[]
```

Comment:
These expected stages come straight from Richard Collete's comment in the JIRA ticket where he suggests what the correct pipeline should look like.

With the exception of some lower/upper case changes it turns out this is exactly what LINQ3 generates.